pkg/dns/plugin.go

```diff
@@ -0,0 +1,284 @@
+
+	h := &IstioServiceEntries{
+		configStore: cfgStore,
+		mux:         dns.NewServeMux(),
+	}
+	h.readServiceEntries(vip)
+	stop := make(chan bool)
+	go func() {
+		ticker := time.NewTicker(5 * time.Second)
+		for {
+			select {
+			case <-stop:
+				return
+			case <-ticker.C:
+				h.readServiceEntries(vip)
+			}
+		}
+	}()
+	h.client = &dns.Client{Net: "udp"}
+	h.server = &dns.Server{Handler: h.mux}
+
+	h.mux.Handle(".", h)
+
+	if grpcServer != nil {
+		// Will be nil on tests, when only subsets of the server are started.
+		dnsapi.RegisterDnsServiceServer(grpcServer, h)
+	}
+	return h
+}
+
+// StartDNS starts the DNS-over-TLS, using a pre-configured listener.
+// Also creates a normal DNS on UDP port 15053, for local debugging
+func (h *IstioServiceEntries) StartDNS(tlsListener net.Listener) {
+	h.server.Listener = tlsListener
+	var err error
+	// UDP - on same port as the MTLS.
+	h.server.PacketConn, err = net.ListenPacket("udp", DNSAddr.Get())
+	if err != nil {
+		log.Warna("Failed to start debugging DNS port", err)
+	}
+	err = h.server.ActivateAndServe()
+	if err != nil {
+		log.Warna("Failed to start debugging DNS", err)
+	}
+}
+
+// Sync the entries from store to the serving map.
+// Should be called on config change ( original code used a 5-sec timer - on the cached entries)
+// TODO: hook it into the config change event.
+func (h *IstioServiceEntries) readServiceEntries(vip string) {
+	dnsEntries := make(map[string][]net.IP)
+	serviceEntries := h.configStore.ServiceEntries()
+
+	for _, e := range serviceEntries {
+		entry := e.Spec.(*networking.ServiceEntry)
+
+		if entry.Resolution == networking.ServiceEntry_NONE {
+			// NO DNS based service discovery for service entries
+			// that specify NONE as the resolution. NONE implies
+			// that Istio should use the IP provided by the caller
+			continue
+		}
+
+		addresses := entry.Addresses
+		if len(addresses) == 0 && vip != "" {
+			// If the ServiceEntry has no Addresses, map to a user-supplied default value, if provided
+			addresses = []string{vip}
+		}
+
+		vips := convertToVIPs(addresses)
+		if len(vips) == 0 {
+			continue
+		}
+
+		for _, host := range entry.Hosts {
+			key := fmt.Sprintf("%s.", host)
+			if strings.Contains(host, "*") {
+				// Validation will ensure that the host is of the form *.foo.com
+				parts := strings.SplitN(host, ".", 2)
+				// Prefix wildcards with a . so that we can distinguish these entries in the map
+				key = fmt.Sprintf(".%s.", parts[1])
+			}
+			dnsEntries[key] = vips
+		}
+	}
+	h.mapMutex.Lock()
+	h.dnsEntries = make(map[string][]net.IP)
+	for k, v := range dnsEntries {
+		h.dnsEntries[k] = v
+	}
+	h.mapMutex.Unlock()
+}
+
+func convertToVIPs(addresses []string) []net.IP {
+	vips := make([]net.IP, 0)
+
+	for _, address := range addresses {
+		// check if its CIDR.  If so, reject the address unless its /32 CIDR
+		if strings.Contains(address, "/") {
```

Comment:
I agree - I tried to not touch the original code too much, and I expect it to change dramatically in separate PRs, since we now have access to the Istiod context - with indexes and other things. 

Since one of the discussions is about having part of it in istio-agent - I'm pretty sure this block will be gone by 1.6 ( and if not - it's only about MC and not a huge problem, I would rather keep it stable)